src/const_eval.md

```diff
@@ -81,11 +81,81 @@ r[const-eval.const-expr.builtin-arith-logic]
   operators used on integer and floating point types, `bool`, and `char`.
 
 r[const-eval.const-expr.borrows]
-* All forms of [borrow]s, including raw borrows, with one limitation:
-  mutable borrows and shared borrows to values with interior mutability
-  are only allowed to refer to *transient* places or to *static* places. A place is *transient*
-  if its lifetime is strictly contained inside the current [const context].
-  A place is *static* if it is a `static` item or a [promoted expression].
+* All forms of [borrow]s, including raw borrows, with one limitation: mutable borrows and shared borrows of expressions whose temporary scope would be extended (see [temporary lifetime extension]) to the end of the program are not allowed when those borrows refer to values with interior mutability.
+
+  ```rust,compile_fail,E0492
+  # use core::sync::atomic::AtomicU8;
+  const C: &AtomicU8 = &AtomicU8::new(0); // ERROR not allowed
+  ```
+
+  ```rust,compile_fail,E0764
+  # use core::sync::atomic::AtomicU8;
+  const C: &AtomicU8 = &mut AtomicU8::new(0); // ERROR not allowed
+  ```
+
+  ```rust,compile_fail,E0492
+  # use core::sync::atomic::AtomicU8;
+  let _: &'static _ = const { &AtomicU8::new(0) }; // ERROR not allowed
+  ```
+
+  ```rust
+  # use core::sync::atomic::AtomicU8;
+  const C: () = { _ = &AtomicU8::new(0); }; // OK
+  ```
+
+  ```rust
+  # use core::sync::atomic::AtomicU8;
+  static S: AtomicU8 = AtomicU8::new(0);
+  const C: &AtomicU8 = &S; // OK
+  ```
+
+  In other words, they are only allowed to refer to *transient* places, to *indirect* places, or to *static* places. A place is *transient* if it is based on a local variable whose lifetime is strictly contained inside the current [const context].
+
+  ```rust
+  const C: () = {
+      let mut x = 0;
+      let _ = &mut x; // Reference to a transient place.
+  };
+  ```
+
+  A place is *indirect* if it is based on a [dereference expression][dereference operator].
+
+  ```rust
+  const C: () = {
+      let x = &mut 0;
+      let _ = &mut *x; // Reference to an indirect place.
+  };
+  ```
+
+  A place is *static* if it is based on a `static` item or a [promoted expression].
+
+  ```rust
+  static mut S: u8 = 0;
+  # #[allow(static_mut_refs)]
+  const C: &u8 = unsafe { &mut S }; // Borrow of static item.
+  ```
+
+  ```rust
+  static mut S: u8 = 0;
+  # #[allow(static_mut_refs)]
+  const C: () = {
+      let _ = unsafe { &mut S }; // Borrow of static item.
+  };
+  ```
+
+  ```rust
+  const C: &[u8] = &mut []; // Borrow of promoted expression.
+  ```
+
+  ```rust
+  const C: &[u8] = &[]; // Borrow of promoted expression.
```

Comment:
This isn't even in the class of "borrows with some form of mutability", so I don't understand how it makes sense as an example here.

Reply:
Nevermind; this needs to be reworked due to https://github.com/rust-lang/reference/pull/1865#discussion_r2172740786 as well.